### PR TITLE
SRL model

### PR DESCRIPTION
**Pull request by DeNeutoy**
- Adds the SRL model.
- Switches the `SimpleTagger` to use `arrays_to_variables`.
- Fixes a bug in `TextField` which meant it didn't return numpy arrays when it wasn't being used within a `Dataset`. 
- Adds viterbi decoding.

2 things which still aren't included here: 
- Masking for the SRL model - I'm planning to do this in another PR for the `SimpleTagger` and `SemanticRoleLabeller` separately, probably using some of the stuff in Matt's PR.
- For inference, we need to inputs to always have a batch dimension, as pytorch doesn't like it when there isn't one. This is a little complex and will require some work on the data api to make it always produce arrays with at least 2 dimensions, so I'll also do that separately.

### CODE REVIEW

**Comment by DeNeutoy** [comment]
Also I could make the binary verb feature into a `BinaryFeatureTextFieldEmbedder` or something - it seemed a bit clunky and not exactly the use case we had originally imagined for the `TextFieldEmbedders`, so I didn't implement it initially.

File: allennlp/common/tensor.py [hunk 1]
```diff
@@ -125,6 +124,56 @@ def masked_softmax(vector, mask):
...
+        for transitioning between a given pair of tags.
+
+    Returns
+    -------
+    viterbi_path : List[int]
+        The tag indices of the maximum likelihood tag sequence.
+    viterbi_score : float
+        The score of the viterbi path.
+    """
+    sequence_length, _ = list(tag_sequence.size())
```
Comment (joelgrus) [inline allennlp/common/tensor.py]: `torch.Size` is a subclass of `tuple`, so you don't need a `list` here

File: allennlp/models/semantic_role_labeller.py [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+                 text_field_embedder: TextFieldEmbedder,
+                 stacked_encoder: Seq2SeqEncoder) -> None:
+        super(SemanticRoleLabeller, self).__init__()
+
+        self.vocab = vocab
+        self.text_field_embedder = text_field_embedder
+        self.num_classes = self.vocab.get_vocab_size("tags")
+
+        # NOTE: You must make sure that the "input_dim" of the stacked encoder
+        # in your configuration file is equal to self.text_field_embedder.output_dim + 1.
```
Comment (joelgrus) [inline allennlp/models/semantic_role_labeller.py]: if this is really required, you should probably check for it and throw a `ConfigurationError`

File: tests/common/tensor_test.py [hunk 3]
```diff
@@ -305,3 +308,41 @@ def test_weighted_sum_handles_uneven_higher_order_input(self):
                                   attention_array[0, i, j, 1] * sentence_array[0, 1])
                 numpy.testing.assert_almost_equal(aggregated_array[0, i, j], expected_array,
                                                   decimal=5)
+
+    def test_viterbi_decode(self):
```
Comment (matt-gardner) [inline tests/common/tensor_test.py]: It'd be nice to have a test against a simple hand-worked example, where you're actually testing the scores, too.

File: allennlp/common/tensor.py [hunk 4]
```diff
@@ -125,6 +124,56 @@ def masked_softmax(vector, mask):
...
+    path_indices = []
+    path_scores.append(tag_sequence[0, :].squeeze())
+
+    # Evaluate the scores for all possible paths.
+    for timestep in range(1, sequence_length):
+        # TODO(Mark): Use broadcasting here once Pytorch 0.2 is released.
+        tiled_path_scores = path_scores[timestep - 1].expand_as(transition_matrix).transpose(0, 1)
+        summed_potentials = tiled_path_scores + transition_matrix
+        scores, paths = torch.max(summed_potentials, 0)
+        path_scores.append(tag_sequence[timestep, :].squeeze() + scores.squeeze())
```
Comment (matt-gardner) [inline allennlp/common/tensor.py]: I don't think you need the squeeze on `tag_sequence` here.

File: allennlp/common/tensor.py [hunk 5]
```diff
@@ -125,6 +124,56 @@ def masked_softmax(vector, mask):
...
+    # Construct the most likely sequence backwards.
+    _, best_path = torch.max(path_scores[-1], 0)
+    viterbi_path = [int(best_path.numpy())]
+    for backward_timestep in reversed(path_indices):
+        viterbi_path.append(int(backward_timestep[viterbi_path[-1]]))
+
+    # Reverse the backward path and find the actual
+    # score of the most likely sequence.
+    viterbi_path.reverse()
+    viterbi_score, _ = torch.max(path_scores[-1], 0)
```
Comment (matt-gardner) [inline allennlp/common/tensor.py]: You can just keep the score from when you get it on line 165.  No need to do the max here again.

File: allennlp/models/semantic_role_labeller.py [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+        expanded_verb_indicator = verb_indicator.unsqueeze(-1).float()
+        # Concatenate the verb feature onto the embedded text. This now
+        # has shape (batch_size, sequence_length, embedding_dim + 1).
+        embedded_text_with_verb_indicator = torch.cat([embedded_text_input, expanded_verb_indicator], -1)
+        batch_size = embedded_text_with_verb_indicator.size()[0]
+        encoded_text = self.stacked_encoder(embedded_text_with_verb_indicator)
+
+        logits = self.tag_projection_layer(encoded_text)
+        reshaped_log_probs = logits.view(-1, self.num_classes)
+        class_probabilities = F.softmax(reshaped_log_probs).view([batch_size, -1, self.num_classes])
```
Comment (matt-gardner) [inline allennlp/models/semantic_role_labeller.py]: I'd recommend making this `.view([batch_size, num_tokens, self.num_classes])`.  That way it will crash here if there's something wrong, instead of failing silently and getting a weird error later.  You can get `num_tokens` up above where you get `batch_size`.

File: allennlp/models/semantic_role_labeller.py [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+        output_dict = {"logits": logits, "class_probabilities": class_probabilities}
```
Comment (matt-gardner) [inline allennlp/models/semantic_role_labeller.py]: The class probabilities aren't mentioned in the docstring as a return value.

File: allennlp/models/semantic_role_labeller.py [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+
+        # NOTE: You must make sure that the "input_dim" of the stacked encoder
+        # in your configuration file is equal to self.text_field_embedder.output_dim + 1.
+        self.stacked_encoder = stacked_encoder
+        self.tag_projection_layer = TimeDistributed(Linear(self.stacked_encoder.get_output_dim(),
+                                                           self.num_classes))
+        # TODO(Mark): support masking once utility functions are merged.
+        self.sequence_loss = torch.nn.CrossEntropyLoss()
+
+    # pylint: disable=arguments-differ
```
Comment (matt-gardner) [inline allennlp/models/semantic_role_labeller.py]: You can just put this right above the docstring, after the method declaration, indented appropriately, and then there's no need to re-enable it, because it should be scoped correctly.  (At least I'm _pretty_ sure that's how pylint works.)

File: allennlp/models/semantic_role_labeller.py [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+                for x in max_likelihood_sequence]
+
+        return {"tags": tags, "class_probabilities": predictions.numpy()}
+
+    def get_viterbi_pairwise_potentials(self):
+        """
+        Generate a matrix of pairwise transition potentials for the BIO labels.
+        The only constraint implemented here is that I-XXX tags must be preceded
+        by either an identical I-XXX tag or a B-XXX tag (i.e they have a pairwise
+        potential of -inf).
```
Comment (matt-gardner) [inline allennlp/models/semantic_role_labeller.py]: Don't you mean "otherwise" instead of "i.e." here?

File: allennlp/models/semantic_role_labeller.py [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+            for j, tag in all_tags.items():
+                # I tags can only be preceded by themselves or
+                # their corresponding B tag.
+                if i != j and tag[0] == 'I' and not previous_tag == 'B' + tag[1:]:
+                    transition_matrix[i, j] = float("-inf")
+        return transition_matrix
+
+    @classmethod
+    def from_params(cls, vocab: Vocabulary, params: Params) -> 'SemanticRoleLabeller':
+        text_field_embedder = TextFieldEmbedder.from_params(vocab, params.pop("text_field_embedder"))
```
Comment (matt-gardner) [inline allennlp/models/semantic_role_labeller.py]: I'd recommend putting in defaults here that replicate Luheng's model, like I did for the BiDAF code.

File: allennlp/models/semantic_role_labeller.py [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+from allennlp.data.fields import IndexField, TextField
+from allennlp.data import Instance
+from allennlp.modules import Seq2SeqEncoder, TimeDistributed, TextFieldEmbedder
+from allennlp.models.model import Model
+
+
+class SemanticRoleLabeller(Model):
+    """
+    This model performs semantic role labelling using BIO tags using Propbank semantic roles.
+    Specifically, it is an implmentation of `Deep Semantic Role Labelling - What works
```
Comment (matt-gardner) [inline allennlp/models/semantic_role_labeller.py]: You changed the spelling of this title...  :us:

File: tests/models/semantic_role_labeller_test.py [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+from allennlp.common.tensor import arrays_to_variables
+from allennlp.data import Vocabulary
+from allennlp.data.dataset_readers import SrlReader
+from allennlp.data.fields import TextField, IndexField
+from allennlp.data.token_indexers import SingleIdTokenIndexer
+from allennlp.models.semantic_role_labeller import SemanticRoleLabeller
+from allennlp.testing.test_case import AllenNlpTestCase
+
+
+class SemanticRoleLabellerTest(AllenNlpTestCase):
```
Comment (matt-gardner) [inline tests/models/semantic_role_labeller_test.py]: Are we testing models save and load?  I noticed on this and the tagger, that test method isn't actually called.

And is it enough for training that `forward` works?  Presumably if `forward` doesn't crash, then backprop and parameter updates will also be fine?

File: allennlp/models/simple_tagger.py [hunk 13]
```diff
@@ -119,11 +120,11 @@ def tag(self, text_field: TextField) -> Dict[str, Any]:
...
         array_input = text_field.as_array(padding_lengths)
-        # TODO(Mark): Generalise how the array is transformed into a variable after settling the data API.
-        # Add a batch dimension by unsqueezing, because pytorch
-        # doesn't support inputs without one.
-        array_input = {"tokens": torch.autograd.Variable(torch.LongTensor(array_input["tokens"])).unsqueeze(0)}
-        output_dict = self.forward(tokens=array_input)
+        # TODO(Mark): Make the data API always return tensors with batch dimensions at every abstraction level.
+        # Add a batch dimension by unsqueezing, because pytorch doesn't support inputs without one.
+        model_input = arrays_to_variables(array_input)
+        model_input["tokens"].data.unsqueeze_(0)
```
Comment (matt-gardner) [inline allennlp/models/simple_tagger.py]: Something like

```python
for input_array in model_input.values():
    input_array.data.unsqueeze_(0)
```
would be more general here.

File: allennlp/models/semantic_role_labeller.py (hunk 9 above)
Comment (DeNeutoy) [inline allennlp/models/semantic_role_labeller.py]: No - the constraint is enforced because they have a `-inf` pairwise potential?

File: allennlp/models/semantic_role_labeller.py (hunk 11 above)
Comment (DeNeutoy) [inline allennlp/models/semantic_role_labeller.py]: I'm lost... what do you want me to spell wrong here?

Comment (matt-gardner) [inline allennlp/models/semantic_role_labeller.py]: "Labelling" is actually "Labeling" in the original paper.  All of these places that you've done "labeller" and "labelling" with two "l"s is a little jarring to my American language model.  I can deal with it, though the quotation of the paper title should probably be correct, so that a copy-paste search turns up the right thing.  Well, Google can deal with the spelling difference, too...

File: allennlp/models/semantic_role_labeller.py (hunk 9 above)
Comment (matt-gardner) [inline allennlp/models/semantic_role_labeller.py]: Yeah, it's just the antecedent of "they" that's unclear.  It looks like it's referring to the `I-XXX` tag and a correct tag preceding it, which would have a potential of 0, while the potential of `-inf` applies to the `I-XXX` tag and _impossible_ preceding tags.

File: allennlp/models/semantic_role_labeller.py (hunk 11 above)
Comment (DeNeutoy) [inline allennlp/models/semantic_role_labeller.py]: Oh right, I hadn't encountered that one before. No problem changing it, just genuinely couldn't see what you meant!

File: tests/models/semantic_role_labeller_test.py (hunk 12 above)
Comment (DeNeutoy) [inline tests/models/semantic_role_labeller_test.py]: Huh. That must have fallen through the cracks of my simple tagger and related PRs. I'll add it.

Regarding your second point, I think so. adding a backward pass to the `test_model_loads_and_saves` function would be easy though, just a case of adding `loss.backward()`. I'll also add this.
